src/main/java/com/google/devtools/build/lib/analysis/constraints/TopLevelConstraintSemantics.java

```diff
@@ -88,6 +105,108 @@ private MissingEnvironment(Label environment, RemovedEnvironmentCulprit culprit)
     }
   }
 
+  /**
+   * Checks that the all top-level targets are compatible with the target platform.
+   *
+   * <p>If any target doesn't support the target platform it will be either marked as "to be
+   * skipped" or marked as "errored".
+   *
+   * <p>Targets that are incompatible with the target platform and are not explicitly requested on
+   * the command line should be skipped.
+   *
+   * <p>Targets that are incompatible with the target platform and *are* explicitly requested on the
+   * command line are errored. Having one or more errored targets will cause the entire build to
+   * fail with an error message.
+   *
+   * @param topLevelTargets the build's top-level targets
+   * @param explicitTargetPatterns the set of explicit target patterns specified by the user on the
+   *     command line.
+   * @return the set of to-be-skipped and errored top-level targets.
+   * @throws ViewCreationFailedException if any top-level target was explicitly requested on the
+   *     command line.
+   */
+  public PlatformRestrictionsResult checkPlatformRestrictions(
+      ImmutableList<ConfiguredTarget> topLevelTargets,
+      ImmutableSet<String> explicitTargetPatterns,
+      boolean keepGoing)
+      throws ViewCreationFailedException {
+    ImmutableSet.Builder<ConfiguredTarget> incompatibleTargets = ImmutableSet.builder();
+    ImmutableSet.Builder<ConfiguredTarget> incompatibleButRequestedTargets = ImmutableSet.builder();
+
+    for (ConfiguredTarget target : topLevelTargets) {
+      IncompatiblePlatformProvider incompatibleProvider =
+          target.getProvider(IncompatiblePlatformProvider.class);
+
+      // Move on to the next target if this one is compatible.
+      if (incompatibleProvider == null) {
+        continue;
+      }
+
+      // We need the label in unambiguous form here. I.e. with the "@" prefix for targets in the
+      // main repository. Otherwise we risk comparing two unrelated targets.
+      String labelString = target.getLabel().getUnambiguousCanonicalForm();
+      if (explicitTargetPatterns.contains(labelString)) {
+        if (!keepGoing) {
+          // Use the slightly simpler form for printing error messages. I.e. no "@" prefix for
+          // targets in the main repository.
+          String targetIncompatibleMessage =
+              ("Target "
```

Comment:
These parens are unneeded. I would re-format this as a template:
private static final String TARGET_INCOMPATIBLE_ERROR_TEMPLATE = "Target %s is incompatible and cannot be built, but was explicitly requested.%s";

String targetIncompatibleMessage = TARGET_INCOMPATIBLE_ERROR_TEMPLATE.format(target.getLabel(),  reportOnIncompatibility());

Reply:
I had to use String.Format(...), but the idea worked.